### ISSUE TITLE
`azd` extension <> Azure Tools don't share credentials

### ISSUE THREAD

**Issue by savannahostrowski** [Azure/azure-dev#1741]
We need to figure out how we can leverage the same token/credentials for the `azd` extension in VS Code as the Azure Tools extension pack.

Right now, this inconsistency means that I can open a project/start a project in VS Code that leverages my work subscription via `azd` but Azure Tools is logged into my personal subscription. As such, I cannot "Show resource group" for my current project until I log out of my personal sub and reauthenticate with my work sub via the Azure Tools extension pack.

![image](https://user-images.githubusercontent.com/8949415/225688516-06a13040-4597-4688-80cf-cc904cbaec49.png)

```[tasklist]
### Tasks
- [x] Define HTTP Based Prototcol to shuttle GetToken requests across processes
- [x] Implement client for protocol in `azd`
- [x] Implement server for protocol in VS Code extension
- [ ] Work through Multi-Tenant Auth Story
- [ ] Make `azd login` do "the right thing" when in integrated auth mode.
```

**Comment by weikanglim** [comment]
This is really interesting. I do agree it's something we need to do short-term.

I wonder what Azure Tools is using under the covers. If it's `az`, I wonder if this is signaling the fact that we have started to diverge away from `az auth` is causing friction.

Long-term, the complex web of "Azure developer tool 1" needs to understand and work with "Azure developer tool 2" and "Azure developer tool 3" in a 3-way graph seems an impossible engineering task, and I think this further solidifies the need for a central auth tool.

**Comment by savannahostrowski** [comment]
I've chatted a bit with @bwateratmsft about this. He'll have the specifics here.

**Comment by esweet431** [comment]
A related ask if we're doing work to unify the auth experience - it would be great to have a shared telemetry property across VS Code Azure events that indicates if the user has authenticated with the Azure extensions/azd. This would enable us to surface different getting started content based on the user's authentication status (i.e., prompt unauthenticated users to sign-in/create an account & nudge authenticated users to become more deeply engaged)

**Comment by ellismg** [comment]
I talked about a prototype solution here with @bwateratmsft that I would like to build.  I'm envisioning a way for `azd` to delegate token fetching to an external process by doing something similar in spirt to MSI where we request a token from a local endpoint (protected by a shared ephemeral key).

We want to prototype something like this during the 1.2.0 timeframe, not committed to shipping it yet, but we want to start getting some investment in some prototypes to show some end to end scenarios.

**Comment by rajeshkamal5050** [comment]
@vhvb1989 moving it back to @ellismg to work with @brandom-msft on this.

**Comment by rajeshkamal5050** [comment]
@ellismg @vhvb1989 @bwateratmsft seeing the POC https://github.com/Azure/azure-dev/pull/2541 has been merged. Can we release this for Nov release 1.5.0?

**Comment by bwateratmsft** [comment]
Should be fine. When do we want to release the ADE / environment changes?

**Comment by rajeshkamal5050** [comment]
@savannahostrowski @ellismg @bwateratmsft Not a must-fix for 1.5.0. Moving it to Jan milestone.

**Comment by bwateratmsft** [comment]
I think @ellismg's [PR](https://github.com/Azure/azure-dev/pull/2541) actually does this though. I don't know if there is any work remaining.

**Comment by savannahostrowski** [comment]
Is this completed and just unreleased then?

**Comment by rajeshkamal5050** [comment]
Remember @vhvb1989 mentioning about the multi-tenant scenarios. May be its out of scope for now and release what we have? Also, should we enable the feature by default in VSCode? @bwateratmsft 

What do you think @ellismg @bwateratmsft

**Comment by bwateratmsft** [comment]
We should be able to run it as an A/B experiment

**Comment by vhvb1989** [comment]
> I think @ellismg's [PR](https://github.com/Azure/azure-dev/pull/2541) actually does this though. I don't know if there is any work remaining.

The PR went in another direction, @bwateratmsft , by having azd using its own authentication within VSCode, and not by re-using the authentication from the azure-account extension.  This was because we were told that the azure-account was going to change and eventually move to do what azd is doing on the PR.  But, we can't release it if the azure-account is not there, as the main story we want to have for customers here is that, if they log in with the azure-account extension, they don't need to log in with azd

**Comment by bwateratmsft** [comment]
The Azure Account extension is eventually going to be deprecated. There is no timeline yet but the current guidance is to use the builtin Microsoft authentication provider.

**Comment by rajeshkamal5050** [comment]
@v-xuto can we test this and make sure it works? 

Currently the feature is [toggled off](https://github.com/Azure/azure-dev/blob/9e05ef2ee47a30179c3d91f428ae8c531e318aea/ext/vscode/package.json#L159). Can be enabled from `vscode azd extension` settings. Refer PR for more details https://github.com/Azure/azure-dev/pull/2541

cc: @gkulin for capturing it in docs FAQ may be?

**Comment by v-jiaodi** [comment]
@rajeshkamal5050 @savannahostrowski  Enabled `azure-dev.auth.useIntegratedAuth` from `vscode azd extension` settings, and try to run `azd auth login` and deploy resource. At this point, my Azure tools display as not logged in and unable to see any resources.
![image](https://github.com/Azure/azure-dev/assets/80496810/dadb7587-8252-4058-888f-d81f7a143612)


When click "Sign in to Azure", we can see the resources created by `azd`.

![image](https://github.com/Azure/azure-dev/assets/80496810/5687847c-8cd6-4101-b370-663cfc8df0ed)

**Comment by rajeshkamal5050** [comment]
> This is expected. There is no way currently for sign in from `azd` to make its way up to VSCode.

@v-jiaodi please validate it the other way around. Post `signing into Azure from VSCode` should not require `signing into azd`

**Comment by v-jiaodi** [comment]
@rajeshkamal5050 When `signing into Azure from VSCode`, you can see Azure Resources, but when executing `azd up`, it prompts: `not logged in, run azd auth login to login`

**Comment by bwateratmsft** [comment]
The exact steps are a little different. The extension launches `azd` processes with a nonce that gives access to the token server. So only processes launched _by the extension_ can benefit from the shared authentication.

1. Enable the setting `azure-dev.auth.useIntegratedAuth`:
![image](https://github.com/Azure/azure-dev/assets/36966225/15e2d617-fc3a-436f-b0d7-a5bf5b17b488)
2. Run `azd auth logout` in the terminal to ensure AZD is not signed in.
3. Run the VSCode command "Azure Developer CLI (azd): Package, Provision and Deploy (up)"
4. You may be prompted to allow the AZD extension to sign in with Microsoft. This should be necessary only once on any given machine. Click "Allow" and follow through with the sign in.
![image](https://github.com/Azure/azure-dev/assets/36966225/4f5a6f85-401b-40e6-98d5-98f4ecc29adf)
5. The extension will run the `azd up` command, this should succeed. It has obtained auth information from the extension.

**Comment by v-jiaodi** [comment]
@bwateratmsft Following the steps you mentioned, it works.
